### PR TITLE
Provide verbose 403 messages explaining related resource problems and other rules

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
Provide a custom message with 403 responses

These messages elaborate our permissions policy. Idea is that every statement could go into documentation, trying to be self-documenting, opening our our comments for users to see, at the time they need them.

##### ISSUE TYPE
 - Feature Pull Request
 - Docs Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION

```

```

##### ADDITIONAL INFORMATION
WIP because this will await some further validation before we can tag to merge.

### CODE REVIEW

**Comment by AlanCoding** [comment]
Demo of functionality in UI:

screen shot 2018-10-12 at 3 56 03 pm

User with admin role to project tries to set its credential to null. Gets 403, which is non-obvious. In our rules, you need permission to existing resource to replace it, so this change surfaces rules like those to the user.

File: awx/api/permissions.py [hunk 1]
```diff
@@ -60,26 +61,31 @@ def check_post_permissions(self, request, view, obj=None):
...
-        else:
-            return check_user_access(request.user, view.model, 'change', obj,
-                                     request.data)
+        has_permission, messages = check_user_access_with_errors(
+            request.user, view.model, 'change', obj, request.data
+        )
+        # Add limited explaination for why an action cannot be taken
+        if messages.get('detail', None):
+            self.message = messages['detail']  # picked up by Django view class
+        return has_permission
```
Comment (AlanCoding) [inline awx/api/permissions.py:88]: This is the right way of doing this.

File: awx/api/views/__init__.py [hunk 2]
```diff
@@ -3682,13 +3683,6 @@ class WorkflowJobRelaunch(WorkflowsEnforcementMixin, GenericAPIView):
     obj_permission_type = 'start'
     serializer_class = EmptySerializer
 
-    def check_object_permissions(self, request, obj):
-        if request.method == 'POST' and obj:
-            relaunch_perm, messages = request.user.can_access_with_errors(self.model, 'start', obj)
-            if not relaunch_perm and 'workflow_job_template' in messages:
-                self.permission_denied(request, message=messages['workflow_job_template'])
-        return super(WorkflowJobRelaunch, self).check_object_permissions(request, obj)
-
```
Comment (AlanCoding) [inline awx/api/views/__init__.py:3150]: This was the wrong way of doing this.

**Comment by AlanCoding** [comment]
The only remaining work I see for this is adding more log messages:

 - editing a sensitive JT field without inventory/project permission
 - only superusers can add instance groups, perhaps shared `only_superusers_can_do_this` method
 - setting `MANAGE_ORGANIZATION_AUTH` prohibits anyone other than superusers from managing users
 - you don't have permission to the object you are associating

etc...

the list could go on and on, but a few high-priority cases would be great to handle.

**Comment by AlanCoding** [comment]
I'm happy with the permission messages now. This should cover the high-priority cases like @benjdudas was recently debugging with the setting for managing permissions.

Taking off the WIP label now, I am ready to start testing / verification for this enhancement.

**Review by wwitzel3 — APPROVED** [review]

**Comment by AlanCoding** [comment]
It probably makes the most sense to delay until the start of the next release cycle, so I will plan on that. However, I am still very interested in merging this. Once that happens, I would like to, and sometime soon I hope to update and resolve conflicts.

**Comment by AlanCoding** [comment]
In light of https://github.com/ansible/awx/pull/3174, I would like to ask @jbradberry if this is a good approach or not.

Our access methods return True or False, except when it causes a permission denied error, in which case it raises an exception. Obviously, raising an exception was on way to convey additional information in a message. Under all other circumstances, though, there is no way to convey additional information to the user, which is sometimes very important. I don't like having 2 different patterns to do this, but I don't see a better way.

**Review by jbradberry — COMMENTED** [review]
@AlanCoding I'm not sure if I'm missing something, but I think the problem is that we weren't using Django's idiom for custom permissions appropriately.  The `False` value isn't supposed to mean "the user definitely does not have permission", but instead that that particular piece of code cannot make the determination and that other checkers downstream should be given the chance.  Raising a `PermissionDenied` error with an appropriate message is supposed to be the correct way of saying definitely no and passing that message back to the user.  I think it would be better to conform to this idiom, adding messages to all of those bare `PermissionDenied` errors and fixing any problematic code that sits in between Django's permission handling code and these individual checkers, though I'd be interested in hearing if you think we have a need that makes it a problem to do it that way.

**Comment by AlanCoding** [comment]
> Raising a PermissionDenied error with an appropriate message is supposed to be the correct way of saying definitely no and passing that message back to the user.

There's two things I want to unpack out of this.

Raising `PermissionDenied` with the text was always the alternative approach I had in mind. This would be a big shakeup, like a _really_ big shakeup, but if that's the more standard approach, then I say we should do it.

I don't know of any case where returning `False` from one of these methods can still result in the user being able to perform the action.

**Comment by jbradberry** [comment]
returning `False`: yeah, I think in the context of what we need we should never need to use that.  `PermissionError` all of the way.

**Comment by AlanCoding** [comment]
we're on the same page. Lot of work to do still.

**Pull request closed without merging**